### PR TITLE
[expo-image][android] Add shared-ref based load API to expo-image

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

It already exists on iOS

# How

Added a load task and are using a sharedRef to get a Glide drawable.

# Test Plan

Unit tests pass

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: apps/test-suite/tests/Image.js [hunk 1]
```diff
@@ -8,7 +8,9 @@ import { mountAndWaitFor, mountAndWaitForWithTimeout, TimeoutError } from './hel
 
 export const name = 'Image';
 
-const REMOTE_SOURCE = { uri: 'https://source.unsplash.com/random' };
+const REMOTE_SOURCE = {
+  uri: 'https://images.unsplash.com/photo-1701743805362-86796f50a0c2?w=1080',
+};
```
Comment (tsapeta) [inline apps/test-suite/tests/Image.js:13]: I know that `/random` endpoint is not working anymore, however now this object is basically the same as `REMOTE_KNOWN_SOURCE` – so maybe a better solution would be to merge it into `REMOTE_SOURCE`?

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt [hunk 2]
```diff
@@ -96,7 +101,37 @@ class ExpoImageModule : Module() {
...
+        // Not relying on `2x` in the filename, but want to make the following true:
+        //  If you multiply the logical size of the image by this value, you get the dimensions of the image in pixels.
+        val screenDensity = appContext.reactContext?.resources?.displayMetrics?.density ?: 1f
+        (image.ref?.toBitmapOrNull()?.density ?: 1) / (screenDensity * 160.0f)
+      }
+      Property("isAnimated") { image: Image ->
+        if (image.ref is GifDrawable) {
+          return@Property true
+        }
+        false
```
Comment (tsapeta) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:127]: What about animated PNG and WebP?

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt [hunk 3]
```diff
@@ -96,7 +101,37 @@ class ExpoImageModule : Module() {
...
+      }
+      Property("mediaType") { ->
+        null // not easily supported on Android https://github.com/bumptech/glide/issues/1378#issuecomment-236879983
```
Comment (tsapeta) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:130]: Can we at least support some of the common image types by inferring them from the file extension?

File: packages/expo-image/src/Image.types.ts [hunk 4]
```diff
@@ -522,12 +522,15 @@ export declare class ImageRef extends SharedRef {
    */
   readonly height: number;
   /**
-   * If you load an image from a file whose name includes the `@2x` modifier, the scale is set to **2.0**.
+   * On iOS, if you load an image from a file whose name includes the `@2x` modifier, the scale is set to **2.0**.
    * All other images are assumed to have a scale factor of **1.0**.
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:526]: I guess a line break is not great now. It seems a bit confusing if it still refers only to iOS or all platforms. Maybe we should reword this?
```suggestion
   * On iOS, if you load an image from a file whose name includes the `@2x` modifier, the scale is set to **2.0**. All other images are assumed to have a scale factor of **1.0**.
```

File: packages/expo-image/src/Image.types.ts [hunk 5]
```diff
@@ -522,12 +522,15 @@ export declare class ImageRef extends SharedRef {
...
-   * If you multiply the logical size of the image by this value, you get the dimensions of the image in pixels.
+   * On Android, it calculates the scale based on the bitmap density divided by screen density.
+   *
+   * On all platforms, if you multiply the logical size of the image by this value, you get the dimensions of the image in pixels.
    */
   readonly scale: number;
   /**
+   * @platform ios
    * Media type (also known as MIME type) of the image, based on its format.
    * Returns `null` when the format is unknown or not supported.
```
Comment (tsapeta) [inline packages/expo-image/src/Image.types.ts:535]:
```suggestion
   * Media type (also known as MIME type) of the image, based on its format.
   * Returns `null` when the format is unknown or not supported.
   * @platform ios
```

**Review by tsapeta — CHANGES_REQUESTED** [review]

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt (hunk 3 above)
Comment (aleqsio) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:130]: We don't really have an extension in a Drawable. I could do that just for animated drawables, maybe @lukmccall knows a better way.

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt (hunk 2 above)
Comment (aleqsio) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:135]: Added

**Review by tsapeta — APPROVED** [review]

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt (hunk 3 above)
Comment (lukmccall) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageModule.kt:138]: We would have to create a custom drawable. We had similar problem decoding SVG - https://github.com/expo/expo/blob/12aeff5b52f9367ee519bd2996f0ea42bc15fcd9/packages/expo-image/android/src/main/java/expo/modules/image/svg/SVGDrawableTranscoder.kt#L19 

Also, we're not passing media type in load event for the same reasons. It's a bigger problem.

File: packages/expo-image/android/src/main/java/expo/modules/image/Image.kt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package expo.modules.image
+
+import android.graphics.drawable.Drawable
+import expo.modules.kotlin.sharedobjects.SharedRef
+
+class Image(ref: Drawable?) : SharedRef<Drawable?>(ref)
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/Image.kt:6]:
```suggestion
class Image(ref: Drawable) : SharedRef<Drawable>(ref)
```
Can we make it not nullable?

Comment (aleqsio) [reply packages/expo-image/android/src/main/java/expo/modules/image/Image.kt]: should be done

File: packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.withContext
+
+open class ImageLoadTask(private val appContext: AppContext, private val source: SourceMap) {
+  private var task: Job? = null
+
+  suspend fun load(promise: Promise?) {
+    return coroutineScope {
+      val deferred = async {
+        val context = this@ImageLoadTask.appContext.reactContext ?: throw Exception("React context is not available")
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt:20]:
```suggestion
        val context = this@ImageLoadTask.appContext.reactContext ?: throw Exceptions.ReactContextLost()
```
You can use common exceptions to handle this case.

File: packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.withContext
+
+open class ImageLoadTask(private val appContext: AppContext, private val source: SourceMap) {
+  private var task: Job? = null
+
+  suspend fun load(promise: Promise?) {
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt:17]:
```suggestion
  suspend fun load(promise: Promise) {
```
It doesn't have to be nullable.

File: packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        withContext(Dispatchers.IO) {
+          Glide.with(context).asDrawable().load(source.uri).submit().get()
+        }
+      }
+      task = deferred
+      try {
+        val bitmap: Drawable = deferred.await()
+        promise?.resolve(Image(bitmap))
+      } catch (e: Exception) {
+        promise?.reject(ImageLoadFailed(e))
```
Comment (lukmccall) [inline packages/expo-image/android/src/main/java/expo/modules/image/ImageLoadTask.kt:30]:
```suggestion
        promise.resolve(Image(bitmap))
      } catch (e: Exception) {
        promise.reject(ImageLoadFailed(e))
```

**Review by lukmccall — APPROVED** [review]

**Pull request merged**